### PR TITLE
[3.12] gh-112302: Backport SBOM generation tooling

### PR DESCRIPTION
**Pull request by sethmlarson**
This PR backports the following PRs to the 3.12 branch which has an identical SBOM to the `main` branch. 3.11 and earlier all have substantial differences (by including `setuptools` which has multiple layers of dependency vendoring). I [proposed backporting to previous branches](https://discuss.python.org/t/create-and-distribute-software-bill-of-materials-sbom-for-python-artifacts/39293/17) on discuss.python.org.

- https://github.com/python/cpython/pull/112303
- https://github.com/python/cpython/pull/112854
- https://github.com/python/cpython/pull/113490
- https://github.com/python/cpython/pull/114450

As a data point to limit fear of breaking builds of downstream distributors, we received feedback from @befeleme in https://github.com/python/cpython/issues/114240 and https://github.com/python/cpython/issues/114244 which appear to have been resolved in https://github.com/python/cpython/pull/114450.

* Issue: gh-112302

### CODE REVIEW

**Review by hugovk — APPROVED** [review]

**Comment by hugovk** [comment]
Does this need updating following https://github.com/python/cpython/pull/115038?

I'm fine with this going in 3.12, do you think it needs RM approval?

**Comment by sethmlarson** [comment]
@hugovk Yes this will need that update applied to this branch, and I indeed want to get approval from release managers, I'll send this again to Thomas today.

**Pull request merged**
